docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,405 @@
+          }
+        }
+      }
+    },
+    "nvidiaGpu": {
+      "compatibilities": {
+        "org.opencontainers.hardware.pci": {
+          "requirements": {
+            "class": "0380",
+            "vendor": "10de"
+          }
+        },
+        "org.opencontainers.kernel.configuration": {
+          "CONFIG_HOTPLUG_PCI_PCIE": "y",
+          "CONFIG_MODULES": "y",
+          "CONFIG_MODULE_UNLOAD": "y",
+          "CONFIG_PCI_MMCONFIG": "y",
+          "CONFIG_DRM_NOUVEAU": "n"
+        }
+      }
+    }
+  }
+}
+```
+
+##### Variant B
+
+Duplicate set of functionaliy from the [Json Graph Specification project](https://github.com/jsongraph/json-graph-specification) for in-house maintainence to ensure complete control and maximum awareness of the specification.
+
+## OCI Compatibility Tool
+
+The compatibility tool maintained by OCI should be very minimal in scope
+and it should export pkgs for external tools so they can cover their own use cases.
+
+The suggested scope of functionality should be as follows:
+
+- Create artifact.
+- Pull and push artifact.
+- Validate compatibility specification.
+- Validate host.
+- Output the compatibility spec in a readable form.
+
+### Implementation
+
+The tool should be written in Golang and implement in-tree plugins architecture.
+
+The structure should be similar to:
+
+```console
+cmd/
+pkg/
+  cmd/
+  plugins/
+    core/
+    community/
+      org.telco/
+      org.supercontainers/
+      ...
+  schema/
+```
+
+- `cmd/` - provided cmds.
+- `pkg/cmd` - all cmds should be written in a form that they can be exported.
+- `pkg/plugins` - plugins extracts and validates information on the host.
+  They must match compatibility domains from the compatibility spec.
+  For instance `org.opencontainers.kernel.module` should be able to list and validate kernel modules on the host.
+- `pkg/plugins/core` - all core plugins maintained by OCI under `org.opencontainers` domain.
+- `pkg/plugins/community` - all community plugins with their own domains with schema.
+  Community maintainers will have access to their own domain directory.
+- `pkg/schema` - schema with graph implementation.
+
+### Core Plugins
+
+The following subjects should be added to core plugins list that are maintained by OCI group:
+
+- CPU
+- Kernel
+- Memory
+- Network
+- PCI
+- USB
+- Storage
+- System
+
+That is duplicated from the [Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery/tree/master/source) project.
+There is a possibility that we could cooperate with NFD community, contribute and import source code responsible for features detection.
+
+### Security
+
+- Linux hosts should be protected by AppArmor/SELinux profiles/policies written by maintainers.
+- Each plugin should have its own section in the profile.
+- The profiles must be carefully reviewed by the maintainers.
+
+The proposal doesn't have a solution for Windows or illumos.
+This can be determined later with Windows and illumos experts.
```

Comment:
There's also WASM, and I'm guessing FreeBSD support will be requested too.

Reply:
Right. FreeBSD was already interested in this project. I think we won't have a problem to incorporate others. We just need some people who knows what to expect and how to implement.